Pay arc changes

cdm-83 opened the pull request:
@jessiagee - Created this new branch and pull request for below changes.

- Added error_code to response object
- Filtered the gateway specific fields from the options 
- Added Credit metho

jessiagee commented:
Hey Senthil,

I had two failures in the `remote_pay_arc_test.rb` file - can you address these and repush once they're fixed?

> Loaded suite test/remote/gateways/remote_pay_arc_test
Started
..........F
========================================================================================================================================================================================================================================
Failure: test_successful_authorize(RemotePayArcTest):
  Response expected to succeed: <#<ActiveMerchant::Billing::MultiResponse:0x00007fc85f911638
   @primary_response=
    #<ActiveMerchant::Billing::Response:0x00007fc89fa3fad8
     @authorization="DoBnOnynynMMMOyW",
     @avs_result=
      {"code"=>nil, "message"=>nil, "street_match"=>nil, "postal_match"=>nil},
     @cvv_result={"code"=>nil, "message"=>nil},
     @emv_authorization=nil,
     @error_code=nil,
     @fraud_review=nil,
     @message=nil,
     @network_transaction_id=nil,
     @params=
      {"data"=>
        {"object"=>"Charge",
         "id"=>"DoBnOnynynMMMOyW",
         "amount"=>100,
         "amount_approved"=>0,
         "amount_refunded"=>0,
         "amount_captured"=>0,
         "amount_voided"=>0,
         "application_fee_amount"=>0,
         "tip_amount"=>0,
         "payarc_fees"=>0,
         "type"=>"Sale",
         "net_amount"=>0,
         "captured"=>0,
         "is_refunded"=>0,
         "status"=>"Bad Request - Try Again",
         "auth_code"=>nil,
         "failure_code"=>"E0200",
         "failure_message"=>"System Error (Tran)",
         "charge_description"=>nil,
         "statement_description"=>"",
         "invoice"=>nil,
         "under_review"=>0,
         "created_at"=>1625067199,
         "updated_at"=>1625067201,
         "email"=>nil,
         "phone_number"=>nil,
         "card_level"=>"LEVEL1",
         "sales_tax"=>nil,
         "purchase_order"=>nil,
         "supplier_reference_number"=>nil,
         "customer_ref_id"=>nil,
         "ship_to_zip"=>nil,
         "amex_descriptor"=>nil,
         "customer_vat_number"=>nil,
         "summary_commodity_code"=>nil,
         "shipping_charges"=>nil,
         "duty_charges"=>nil,
         "ship_from_zip"=>nil,
         "destination_country_code"=>nil,
         "vat_invoice"=>nil,
         "order_date"=>nil,
         "tax_category"=>nil,
         "tax_type"=>nil,
         "tax_rate"=>nil,
         "tax_amount"=>nil,
         "created_by"=>"senthil.j@trackdfect.com",
         "terminal_register"=>nil,
         "tip_amount_refunded"=>nil,
         "sales_tax_refunded"=>nil,
         "shipping_charges_refunded"=>nil,
         "duty_charges_refunded"=>nil,
         "pax_reference_number"=>nil,
         "refund_reason"=>nil,
         "refund_description"=>nil,
         "surcharge"=>0,
         "toll_amount"=>nil,
         "refund"=>{"data"=>[]},
         "card"=>
          {"data"=>
            {"object"=>"Card",
             "id"=>"15y2901NNL550MLv",
             "address1"=>"920 Sunnyslope Ave",
             "address2"=>"Bronx",
             "card_source"=>"INTERNET",
             "card_holder_name"=>"Rex Joseph",
             "is_default"=>0,
             "exp_month"=>"12",
             "exp_year"=>"2022",
             "is_verified"=>1,
             "fingerprint"=>"1Lv0NN9LyN5Pm105",
             "city"=>"New York",
             "state"=>"New York",
             "zip"=>"10469",
             "brand"=>"V",
             "last4digit"=>"1111",
             "first6digit"=>411111,
             "country"=>"USA",
             "avs_status"=>"N",
             "cvc_status"=>"M",
             "address_check_passed"=>0,
             "zip_check_passed"=>0,
             "customer_id"=>nil,
             "created_at"=>1625067199,
             "updated_at"=>1625067201}}},
       "meta"=>{"include"=>["review"], "custom"=>[]}},
     @success=false,
     @test=true>,
   @responses=
    [#<ActiveMerchant::Billing::Response:0x00007fc85f9025e8
      @authorization="ELNm8w0LmqN8w0lq",
      @avs_result=
       {"code"=>nil, "message"=>nil, "street_match"=>nil, "postal_match"=>nil},
      @cvv_result={"code"=>nil, "message"=>nil},
      @emv_authorization=nil,
      @error_code=nil,
      @fraud_review=nil,
      @message="ELNm8w0LmqN8w0lq",
      @network_transaction_id=nil,
      @params=
       {"data"=>
         {"object"=>"Token",
          "id"=>"ELNm8w0LmqN8w0lq",
          "used"=>false,
          "ip"=>nil,
          "tokenization_method"=>nil,
          "created_at"=>1625067197,
          "updated_at"=>1625067199,
          "card"=>
           {"data"=>
             {"object"=>"Card",
              "id"=>"15y2901NNL550MLv",
              "address1"=>"920 Sunnyslope Ave",
              "address2"=>"Bronx",
              "card_source"=>"INTERNET",
              "card_holder_name"=>"Rex Joseph",
              "is_default"=>0,
              "exp_month"=>"12",
              "exp_year"=>"2022",
              "is_verified"=>1,
              "fingerprint"=>"1Lv0NN9LyN5Pm105",
              "city"=>"New York",
              "state"=>"New York",
              "zip"=>"10469",
              "brand"=>"V",
              "last4digit"=>"1111",
              "first6digit"=>411111,
              "country"=>"USA",
              "avs_status"=>"N",
              "cvc_status"=>"M",
              "address_check_passed"=>0,
              "zip_check_passed"=>0,
              "customer_id"=>nil,
              "created_at"=>1625067199,
              "updated_at"=>1625067199}}},
        "meta"=>{"include"=>[], "custom"=>[]}},
      @success=true,
      @test=true>,
     #<ActiveMerchant::Billing::Response:0x00007fc89fa3fad8
      @authorization="DoBnOnynynMMMOyW",
      @avs_result=
       {"code"=>nil, "message"=>nil, "street_match"=>nil, "postal_match"=>nil},
      @cvv_result={"code"=>nil, "message"=>nil},
      @emv_authorization=nil,
      @error_code=nil,
      @fraud_review=nil,
      @message=nil,
      @network_transaction_id=nil,
      @params=
       {"data"=>
         {"object"=>"Charge",
          "id"=>"DoBnOnynynMMMOyW",
          "amount"=>100,
          "amount_approved"=>0,
          "amount_refunded"=>0,
          "amount_captured"=>0,
          "amount_voided"=>0,
          "application_fee_amount"=>0,
          "tip_amount"=>0,
          "payarc_fees"=>0,
          "type"=>"Sale",
          "net_amount"=>0,
          "captured"=>0,
          "is_refunded"=>0,
          "status"=>"Bad Request - Try Again",
          "auth_code"=>nil,
          "failure_code"=>"E0200",
          "failure_message"=>"System Error (Tran)",
          "charge_description"=>nil,
          "statement_description"=>"",
          "invoice"=>nil,
          "under_review"=>0,
          "created_at"=>1625067199,
          "updated_at"=>1625067201,
          "email"=>nil,
          "phone_number"=>nil,
          "card_level"=>"LEVEL1",
          "sales_tax"=>nil,
          "purchase_order"=>nil,
          "supplier_reference_number"=>nil,
          "customer_ref_id"=>nil,
          "ship_to_zip"=>nil,
          "amex_descriptor"=>nil,
          "customer_vat_number"=>nil,
          "summary_commodity_code"=>nil,
          "shipping_charges"=>nil,
          "duty_charges"=>nil,
          "ship_from_zip"=>nil,
          "destination_country_code"=>nil,
          "vat_invoice"=>nil,
          "order_date"=>nil,
          "tax_category"=>nil,
          "tax_type"=>nil,
          "tax_rate"=>nil,
          "tax_amount"=>nil,
          "created_by"=>"senthil.j@trackdfect.com",
          "terminal_register"=>nil,
          "tip_amount_refunded"=>nil,
          "sales_tax_refunded"=>nil,
          "shipping_charges_refunded"=>nil,
          "duty_charges_refunded"=>nil,
          "pax_reference_number"=>nil,
          "refund_reason"=>nil,
          "refund_description"=>nil,
          "surcharge"=>0,
          "toll_amount"=>nil,
          "refund"=>{"data"=>[]},
          "card"=>
           {"data"=>
             {"object"=>"Card",
              "id"=>"15y2901NNL550MLv",
              "address1"=>"920 Sunnyslope Ave",
              "address2"=>"Bronx",
              "card_source"=>"INTERNET",
              "card_holder_name"=>"Rex Joseph",
              "is_default"=>0,
              "exp_month"=>"12",
              "exp_year"=>"2022",
              "is_verified"=>1,
              "fingerprint"=>"1Lv0NN9LyN5Pm105",
              "city"=>"New York",
              "state"=>"New York",
              "zip"=>"10469",
              "brand"=>"V",
              "last4digit"=>"1111",
              "first6digit"=>411111,
              "country"=>"USA",
              "avs_status"=>"N",
              "cvc_status"=>"M",
              "address_check_passed"=>0,
              "zip_check_passed"=>0,
              "customer_id"=>nil,
              "created_at"=>1625067199,
              "updated_at"=>1625067201}}},
        "meta"=>{"include"=>["review"], "custom"=>[]}},
      @success=false,
      @test=true>],
   @use_first_response=false>>
test/remote/gateways/remote_pay_arc_test.rb:62:in `test_successful_authorize'
     59: 
     60:   def test_successful_authorize
     61:     response = @gateway.authorize(@amount, @credit_card, @options)
  => 62:     assert_success response
     63:     assert_equal 'authorized', response.message
     64:   end
     65: 
========================================================================================================================================================================================================================================
.F
========================================================================================================================================================================================================================================
Failure: test_successful_credit(RemotePayArcTest):
  Response expected to succeed: <#<ActiveMerchant::Billing::Response:0x00007fc89f42d758
   @authorization=nil,
   @avs_result=
    {"code"=>nil, "message"=>nil, "street_match"=>nil, "postal_match"=>nil},
   @cvv_result={"code"=>nil, "message"=>nil},
   @emv_authorization=nil,
   @error_code=403,
   @fraud_review=nil,
   @message="Account Setting is not set up for blind credit",
   @network_transaction_id=nil,
   @params=
    {"status"=>"error",
     "code"=>0,
     "message"=>"Account Setting is not set up for blind credit",
     "status_code"=>403,
     "exception"=>"App\\Ship\\Exceptions\\InvalidAccountException",
     "file"=>
      "/home/deploy/payarc.com/app/Containers/Refund/Actions/CreateRefundWoReferenceAction.php",
     "line"=>122,
     "trace"=>
      "App\\Ship\\Exceptions\\InvalidAccountException: Account Setting is not set up for blind credit in /home/deploy/payarc.com/app/Containers/Refund/Actions/CreateRefundWoReferenceAction.php:122\n" +
      "Stack trace:\n" +
      "#0 /home/deploy/payarc.com/app/Containers/Refund/Actions/CreateRefundWoReferenceAction.php(46): App\\Containers\\Refund\\Actions\\CreateRefundWoReferenceAction->validateData()\n" +
      "#1 /home/deploy/payarc.com/apiato/core/Traits/CallableTrait.php(43): App\\Containers\\Refund\\Actions\\CreateRefundWoReferenceAction->run(Object(App\\Containers\\Refund\\UI\\API\\Requests\\CreateRefundWoReferenceRequest))\n" +
      "#2 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(239): Apiato\\Core\\Foundation\\Apiato->call(Object(App\\Containers\\Refund\\Actions\\CreateRefundWoReferenceAction), Array)\n" +
      "#3 /home/deploy/payarc.com/app/Containers/Refund/UI/API/Controllers/Controller.php(106): Illuminate\\Support\\Facades\\Facade::__callStatic('call', Array)\n" +
      "#4 [internal function]: App\\Containers\\Refund\\UI\\API\\Controllers\\Controller->createRefundWoReference(Object(App\\Containers\\Refund\\UI\\API\\Requests\\CreateRefundWoReferenceRequest))\n" +
      "#5 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): call_user_func_array(Array, Array)\n" +
      "#6 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(45): Illuminate\\Routing\\Controller->callAction('createRefundWoR...', Array)\n" +
      "#7 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Route.php(219): Illuminate\\Routing\\ControllerDispatcher->dispatch(Object(Illuminate\\Routing\\Route), Object(App\\Containers\\Refund\\UI\\API\\Controllers\\Controller), 'createRefundWoR...')\n" +
      "#8 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Route.php(176): Illuminate\\Routing\\Route->runController()\n" +
      "#9 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(680): Illuminate\\Routing\\Route->run()\n" +
      "#10 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#11 /home/deploy/payarc.com/app/Containers/Localization/Middlewares/LocalizationMiddleware.php(37): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#12 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): App\\Containers\\Localization\\Middlewares\\LocalizationMiddleware->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#13 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#14 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#15 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#16 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#17 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(66): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#18 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#19 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#20 /home/deploy/payarc.com/app/Ship/Middlewares/Http/StartSession.php(70): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#21 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): App\\Ship\\Middlewares\\Http\\StartSession->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#22 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#23 /home/deploy/payarc.com/app/Ship/Middlewares/Http/ProfilerMiddleware.php(26): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#24 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): App\\Ship\\Middlewares\\Http\\ProfilerMiddleware->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#25 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#26 /home/deploy/payarc.com/app/Ship/Middlewares/Http/ProcessETagHeadersMiddleware.php(38): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#27 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): App\\Ship\\Middlewares\\Http\\ProcessETagHeadersMiddleware->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#28 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#29 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(41): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#30 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#31 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#32 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Auth/Middleware/Authenticate.php(43): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#33 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Auth\\Middleware\\Authenticate->handle(Object(Illuminate\\Http\\Request), Object(Closure), 'api')\n" +
      "#34 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#35 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#36 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(682): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))\n" +
      "#37 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(657): Illuminate\\Routing\\Router->runRouteWithinStack(Object(Illuminate\\Routing\\Route), Object(Illuminate\\Http\\Request))\n" +
      "#38 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(623): Illuminate\\Routing\\Router->runRoute(Object(Illuminate\\Http\\Request), Object(Illuminate\\Routing\\Route))\n" +
      "#39 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(612): Illuminate\\Routing\\Router->dispatchToRoute(Object(Illuminate\\Http\\Request))\n" +
      "#40 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(176): Illuminate\\Routing\\Router->dispatch(Object(Illuminate\\Http\\Request))\n" +
      "#41 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#42 /home/deploy/payarc.com/vendor/laravel/nova/src/Http/Middleware/ServeNova.php(26): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#43 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Laravel\\Nova\\Http\\Middleware\\ServeNova->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#44 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#45 /home/deploy/payarc.com/vendor/barryvdh/laravel-cors/src/HandleCors.php(36): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#46 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Barryvdh\\Cors\\HandleCors->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#47 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#48 /home/deploy/payarc.com/app/Ship/Middlewares/RequestLoggerMiddleware.php(40): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#49 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): App\\Ship\\Middlewares\\RequestLoggerMiddleware->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#50 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#51 /home/deploy/payarc.com/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#52 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Fideloper\\Proxy\\TrustProxies->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#53 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#54 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#55 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#56 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#57 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#58 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#59 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#60 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#61 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#62 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#63 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(62): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#64 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Foundation\\Http\\Middleware\\CheckForMaintenanceMode->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#65 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#66 /home/deploy/payarc.com/app/Containers/Debugger/Middlewares/RequestsMonitorMiddleware.php(27): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#67 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): App\\Containers\\Debugger\\Middlewares\\RequestsMonitorMiddleware->handle(Object(Illuminate\\Http\\Request), Object(Closure))\n" +
      "#68 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#69 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))\n" +
      "#70 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(151): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))\n" +
      "#71 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(116): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter(Object(Illuminate\\Http\\Request))\n" +
      "#72 /home/deploy/payarc.com/public/index.php(54): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))\n" +
      "#73 {main}"},
   @success=false,
   @test=true>>
test/remote/gateways/remote_pay_arc_test.rb:149:in `test_successful_credit'
     146: 
     147:   def test_successful_credit
     148:     response = @gateway.credit(@amount, @credit_card, @options)
  => 149:     assert_success response
     150:     assert_block do
     151:       PayArcGateway::SUCCESS_STATUS.include? response.message
     152:     end
========================================================================================================================================================================================================================================
......
Finished in 61.328207 seconds.
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
19 tests, 45 assertions, 2 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
89.4737% passed
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
0.31 tests/s, 0.73 assertions/s

jessiagee reviewed (changes requested):
Overall the changes look good aside from the two failing remote tests I posted below.

Please add a remote test for the `error_code_from` addition to ensure that we are getting back error codes as we would expect them.

Thank you!

cdm-83 commented:
> Overall the changes look good aside from the two failing remote tests I posted below.
> 
> Please add a remote test for the `error_code_from` addition to ensure that we are getting back error codes as we would expect them.
> 
> Thank you!

Hi @jessiagee ,

Added error_code  check in few test cases. 
The test case failure is due to some issues in server is resolving, otherwise there will not be any change in the code.

Senthil

jessiagee commented:
> > Overall the changes look good aside from the two failing remote tests I posted below.
> > Please add a remote test for the `error_code_from` addition to ensure that we are getting back error codes as we would expect them.
> > Thank you!
> 
> Hi @jessiagee ,
> 
> Added error_code check in few test cases.
> The test case failure is due to some issues in server is resolving, otherwise there will not be any change in the code.
> 
> Senthil

Hey Senthil,

What will we need to do to fix the issue with the server so that we can get those tests to pass? Is that on PayArc's side or do we need to get something set on our account so that we can have these tests pass?

Please make sure to merge, rebase, and squash, and then I can approve the PR.

Thanks,
Jessi

jessiagee commented:
@cdm-83 

I pulled these changes over to test the Spreedly integration and am getting the following error on authorize, purchase, capture, refund, and void.

```{\\\"status\\\":\\\"error\\\",\\\"code\\\":0,\\\"message\\\":\\\"The
  token id and customer id field cannot be present together.Only one needs to be specified.\\\",\\\"status_code\\\":500,\\\"exception\\\":\\\"Symfony\\\\\\\\Component\\\\\\\\Routing\\\\\\\\Exception\\\\\\\\InvalidParameterException\```

cdm-83 commented:
> @cdm-83
> 
> I pulled these changes over to test the Spreedly integration and am getting the following error on authorize, purchase, capture, refund, and void.
> 
> ```
>   token id and customer id field cannot be present together.Only one needs to be specified.\\\",\\\"status_code\\\":500,\\\"exception\\\":\\\"Symfony\\\\\\\\Component\\\\\\\\Routing\\\\\\\\Exception\\\\\\\\InvalidParameterException\```
> ```

@jessiagee  - Checked with Punith,  token id, and customer id both cannot be passed together. Only one can be passed at a time.  If Spreedly has a scenario to pass both then the API has to support, API-related queries you can directly check with Punith.  Anything related to the extension of Active Merchant I can help.

Thanks!
Senthil

cdm-83 commented:
> > > Overall the changes look good aside from the two failing remote tests I posted below.
> > > Please add a remote test for the `error_code_from` addition to ensure that we are getting back error codes as we would expect them.
> > > Thank you!
> > 
> > 
> > Hi @jessiagee ,
> > Added error_code check in few test cases.
> > The test case failure is due to some issues in server is resolving, otherwise there will not be any change in the code.
> > Senthil
> 
> Hey Senthil,
> 
> What will we need to do to fix the issue with the server so that we can get those tests to pass? Is that on PayArc's side or do we need to get something set on our account so that we can have these tests pass?
> 
> Once you've rebased and squashed, I'll approve and merge in the PR.
> 
> Thanks,
> Jessi

Hi @jessiagee 
Those two est cases are passing now, but another test is a case related to Credit is failing, API team is fixing. 
I have rebased and squashed it into a single commit.  I don't see any code change for this failing case, so if you are fine with everything can you merge with PR ?

Thanks
Senthil

cdm-83 commented:
@jessiagee - All the test cases are passing now and code is also squashed and pushed. 
Please have a look. 

Thanks,
Senthil

jessiagee commented:
> @jessiagee - All the test cases are passing now and code is also squashed and pushed.
> Please have a look.
> 
> Thanks,
> Senthil

Hi Senthil, 

The office was closed yesterday for a holiday so looking at the PR now.

I'm still seeing two remote test failures with `test_successful_credit` and `test_successful_authorize_and_capture`:

> Failure: test_successful_credit(RemotePayArcTest):
  Response expected to succeed: <#<ActiveMerchant::Billing::Response:0x00007f9468ab3368
   @authorization=nil,
   @avs_result=
    {"code"=>nil, "message"=>nil, "street_match"=>nil, "postal_match"=>nil},
   @cvv_result={"code"=>nil, "message"=>nil},
   @emv_authorization=nil,
   @error_code=403,
   @fraud_review=nil,
   @message="Account Setting is not set up for blind credit",
   @network_transaction_id=nil,
   @params=
    {"status"=>"error",
     "code"=>0,
     "message"=>"Account Setting is not set up for blind credit",
     "status_code"=>403,
     "exception"=>"App\\Ship\\Exceptions\\InvalidAccountException",
     "file"=>
      "/home/deploy/payarc.com/app/Containers/Refund/Actions/CreateRefundWoReferenceAction.php",
     "line"=>122,
     "trace"=>
      "App\\Ship\\Exceptions\\InvalidAccountException: Account Setting is not set up for blind credit in /home/deploy/payarc.com/app/Containers/Refund/Actions/CreateRefundWoReferenceAction.php:122\n" +
      "Stack trace:\n" +
      "#0 /home/deploy/payarc.com/app/Containers/Refund/Actions/CreateRefundWoReferenceAction.php(46): App\\Containers\\Refund\\Actions\\CreateRefundWoReferenceAction->validateData()\n" +
      "#1 /home/deploy/payarc.com/apiato/core/Traits/CallableTrait.php(43): App\\Containers\\Refund\\Actions\\CreateRefundWoReferenceAction->run(Object(App\\Containers\\Refund\\UI\\API\\Requests\\CreateRefundWoReferenceRequest))\n" +
      "#2 /home/deploy/payarc.com/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(239): Apiato\\Core\\Foundation\\Apiato->call(Object(App\\Containers\\Refund\\Actions\\CreateRefundWoReferenceAction), Array)\n" +

> Failure: test_successful_authorize_and_capture(RemotePayArcTest):
  Response expected to succeed: <#<ActiveMerchant::Billing::MultiResponse:0x00007f94689e6368
   @primary_response=
    #<ActiveMerchant::Billing::Response:0x00007f9468a12bc0
     @authorization="XMbnObRnoboLXORo",
     @avs_result=
      {"code"=>nil, "message"=>nil, "street_match"=>nil, "postal_match"=>nil},
     @cvv_result={"code"=>nil, "message"=>nil},
     @emv_authorization=nil,
     @error_code=nil,
     @fraud_review=nil,
     @message=nil,
     @network_transaction_id=nil,
     @params=
      {"data"=>
        {"object"=>"Charge",
         "id"=>"XMbnObRnoboLXORo",
         "amount"=>100,
         "amount_approved"=>0,
         "amount_refunded"=>0,
         "amount_captured"=>0,
         "amount_voided"=>0,
         "application_fee_amount"=>0,
         "tip_amount"=>0,
         "payarc_fees"=>0,
         "type"=>"Sale",
         "net_amount"=>0,
         "captured"=>0,
         "is_refunded"=>0,
         "status"=>"failed_by_gateway",
         "auth_code"=>nil,
         "failure_code"=>"D0008",
         "failure_message"=>"Possible Duplicate Request",

Can you check with the API team and let me know what's happening here?

Thanks,
Jessi

cdm-83 commented:
Hi @jessiagee ,

It should be fine now, All cases are passing for me.

...................
Finished in 109.310173658 seconds.
--------------------------------------------------------------------------------
19 tests, 50 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
--------------------------------------------------------------------------------

Senthil

jessiagee commented:
> Hi @jessiagee ,
> 
> It should be fine now, All cases are passing for me.
> 
> ## ...................
> Finished in 109.310173658 seconds.
> ## 19 tests, 50 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
> 100% passed
> Senthil

Hey Senthil,

I reached out to Punith at PayArc about these failures. The `test_successful_credit` one should be related to needing to flip a setting on our accounts which we are working to get done.

Punith's response for `test_successful_authorize_and_capture` is here: "[This is] throwing an error "possible duplicate request". This happens when you run a transaction for a merchant using the same card and the same amount. For your test case, you can simply change the amount and run it again."

I changed the amounts listed in `test_successful_authorize_and_capture` to be `2000` instead of `@amount` and got a successful response. Please update that test so that others won't potentially run into it, and then I can go ahead and approve/merge the PR.

Thanks!
Jessi

jessiagee reviewed (changes requested):
Added comment above for changes

cdm-83 commented:
Hi @jessiagee,

Changed amount values in the test case and all cases are passing for me, Please check. 

Senthil

jessiagee reviewed (approved):
Approving this though there is a rubocop issue that I'll need to fix as well
